### PR TITLE
add version check for existing packages

### PR DESCRIPTION
**Pull request by juanvallejo**
Addresses https://github.com/openshift/openshift-ansible/pull/4064#pullrequestreview-35997772

Adds a version check for existing versions of Open vSwitch

TODO

- [x] add tests

cc @sosiouxme @brenton @rhcarvalho

### CODE REVIEW

**Comment by rhcarvalho** [comment]
A PR to address an open PR? That's new :-)

File: roles/openshift_health_checker/openshift_checks/existing_ovs_version.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
+# pylint: disable=missing-docstring
```
Comment (rhcarvalho) [inline roles/openshift_health_checker/openshift_checks/existing_ovs_version.py]: @juanvallejo please let's replace this with docstrings :-)

File: roles/openshift_health_checker/openshift_checks/existing_ovs_version.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+"""
+Ansible module for determining if an installed version of Open vSwitch is incompatible with the
+currently installed version of OpenShift.
+"""
+
+from openshift_checks import OpenShiftCheck, OpenShiftCheckException, get_var
+
+
+class ExistingOvsVersion(OpenShiftCheck):
```
Comment (rhcarvalho) [inline roles/openshift_health_checker/openshift_checks/existing_ovs_version.py]: Missing docstring

File: roles/openshift_health_checker/openshift_checks/existing_ovs_version.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+    name = "existing_ovs_version"
```
Comment (rhcarvalho) [inline roles/openshift_health_checker/openshift_checks/existing_ovs_version.py]: Naming is hard. Wouldn't "ovs_version" be enough?

File: roles/openshift_health_checker/openshift_checks/existing_ovs_version.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+                    self.get_required_ovs_version(task_vars),
+                ),
+            ],
+        }
+        return self.execute_module("rpm_version", args, task_vars)
+
+    def get_required_ovs_version(self, task_vars):
+        """Return the correct Open vSwitch version for the current OpenShift version.
+        If the current OpenShift version is >= 3.5, ensure Open vSwitch version 2.6,
+        Else ensure Open vSwitch version 2.4"""
```
Comment (rhcarvalho) [inline roles/openshift_health_checker/openshift_checks/existing_ovs_version.py]: We don't need to describe the implementation details in a docstring. As soon as the implementation changes, the documentation would easily become wrong.

File: roles/openshift_health_checker/library/rpm_version.py [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+
+    if IMPORT_EXCEPTION:
+        module.fail_json(msg="rpm_version module could not import rpm: %s" % IMPORT_EXCEPTION)
+
+    # determine the packages we will look for
+    pkg_list = module.params['package_list']
+    if not pkg_list:
+        module.fail_json(msg="package_list must not be empty")
+
+    # get the list of packages available and complain if any are
```
Comment (ingvagabund) [inline roles/openshift_health_checker/library/rpm_version.py]: s/if any are/if any of them is/

File: roles/openshift_health_checker/library/rpm_version.py [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+    # missing or any errors occur
+    try:
+        pkg_dict = {pkg[LIST_PKG_NAME]: pkg for pkg in pkg_list}
```
Comment (ingvagabund) [inline roles/openshift_health_checker/library/rpm_version.py]: Would be great to share an example of expected usage. E.g. is it
```
rpm_version:
  package_list:
    - 0: 4.3.42-7
      1: bash
```

or 

```
rpm_version:
  package_list:
    - version: 4.3.42-7
      name: bash
```

?

Comment (juanvallejo) [inline roles/openshift_health_checker/library/rpm_version.py]: @ingvagabund each item under `package_list` is a list consisting of two items (installed_package_name, required_version).

```
--- 
rpm_version: 
  package_list: 
    - 
      - pkg_name
      - "1.2.3"
    - 
      - pkg_name_2
      - "4.5.6"
```

**Comment by juanvallejo** [comment]
@rhcarvalho @ingvagabund thanks for the feedback, will begin adding tests

File: roles/openshift_health_checker/library/rpm_version.py [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+def _to_dict(pkg_list):
+    return {pkg[LIST_PKG_NAME]: pkg for pkg in pkg_list}
+
+
+# pylint: disable=invalid-name
+def _retrieve_expected_package_versions(expected_pkgs_dict):
+    """Search for installed packages matching given pkg names
+    and versions. Returns a dictionary: {pkg_name: [versions]}"""
+
+    ts = rpm.TransactionSet()  # pylint: disable=invalid-name
```
Comment (sosiouxme) [inline roles/openshift_health_checker/library/rpm_version.py]: https://github.com/openshift/openshift-ansible/blob/master/docs/best_practices_guide.adoc#All-PyLint-rule-disables-MUST-be-documented-in-the-code

**Comment by juanvallejo** [comment]
@sosiouxme thanks for the feedback, also addressed additional review comments from https://github.com/openshift/openshift-ansible/pull/4064 that were also applicable to this PR

**Comment by juanvallejo** [comment]
aos-ci-test

**Comment by sosiouxme** [comment]
[merge]
(and for once I actually checked locally that integration tests pass)

**Comment by sosiouxme** [comment]
merge failed by missing devcut... try again when master reopens

**Comment by sosiouxme** [comment]
[merge]

**Comment by sosiouxme** [comment]
@openshift/team-project-committers do you want devs to bug you to rerun merges when the only reason it failed last time is because it was post-devcut? Or can the bot be smarter...

Also please start the merge for us here and on https://github.com/openshift/openshift-ansible/pull/4064

**Comment by bparees** [comment]
[merge]

**Comment by bparees** [comment]
> @openshift/team-project-committers do you want devs to bug you to rerun merges when the only reason it failed last time is because it was post-devcut? Or can the bot be smarter...

@stevekuznetsov it would be nice if the bot could be smarter...  how hard is it?

**Pull request merged**